bilge-impl/src/bitsize/split.rs

```diff
@@ -0,0 +1,185 @@
+/// the whole family of `-Bits` macros only works when used in that scope.
+/// 
+/// Let's visualize why this is the case, starting with some user-code:
+/// ```ignore
+/// #[bitsize(6)]
+/// #[derive(Clone, Copy, PartialEq, DebugBits, FromBits)]
+/// struct Example {
+///     field1: u2,
+///     field2: u4,
+/// }
+/// ```
+/// First, the attributes get sorted, depending on their name.
+/// Every attribute in need of field information gets resolved first,
+/// in this case `DebugBits` and `FromBits`.
+/// 
+/// Now, after resolving all `before_compression` attributes, the halfway-resolved
+/// code looks like this:
+/// ```ignore
+/// #[bilge::bitsize_internal(6)]
+/// #[derive(Clone, Copy, PartialEq)]
+/// struct Example {
+///     field1: u2,
+///     field2: u4,
+/// }
+/// ```
+/// This `#[bitsize_internal]` attribute is the one actually doing the compression and generating
+/// all the getters, setters and a constructor.
+/// 
+/// Finally, the struct ends up like this (excluding the generated impl blocks):
+/// ```ignore
+/// struct Example {
+///     value: u6,
+/// }
+/// ```
+pub struct SplitAttributes {
+    pub before_compression: Vec<Attribute>,
+    pub after_compression: Vec<Attribute>,
+}
+
+/// Split item attributes into those applied before bitfield-compression and those applied after.
+/// Also, abort on any invalid configuration.
+/// 
+/// Any derives with suffix `Bits` will be able to access field information.
+/// This way, users of `bilge` can define their own derives working on the uncompressed bitfield.
+pub fn split_item_attributes(item: &Item) -> SplitAttributes {
+    let attrs = match item {
+        Item::Enum(item) => &item.attrs,
+        Item::Struct(item) => &item.attrs,
+        _ => abort_call_site!("item is not a struct or enum"; help = "`#[bitsize]` can only be used on structs and enums"),
+    };
+
+    let parsed = attrs.iter().map(parse_attribute);
+    
+    let is_struct = matches!(item, Item::Struct(..));
+    
+    let mut from_bytes = None;
+    let mut has_frombits = false;
+
+    let mut before_compression = vec![];
+    let mut after_compression = vec![];
+
+    for parsed_attr in parsed {
+        match parsed_attr {
+            ParsedAttribute::DeriveList(derives) => {
+                for derive in derives {
+                    // NOTE: we could also handle `::{path}`
+                    match derive.to_string().as_str() {
+                        "FromBytes" | "zerocopy :: FromBytes" => from_bytes = Some(derive.clone()),
+                        "FromBits" | "bilge :: FromBits" => has_frombits = true,
+                        "Debug" | "fmt :: Debug" | "core :: fmt :: Debug" | "std :: fmt :: Debug" if is_struct => {
+                            abort!(derive.0, "use derive(DebugBits) for structs")
+                        }
+                        _ => {}
+                    };
+
+                
+                    if derive.is_bitfield_derive() {
```

Comment:
addressing the TODO below:
As this is a main feature, something like `is_custom_bitfield_derive`?